paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/RowDataChannelComputer.java

```diff
@@ -18,20 +18,54 @@
 
 package org.apache.paimon.flink.sink;
 
+import org.apache.paimon.data.BinaryRow;
 import org.apache.paimon.schema.TableSchema;
+import org.apache.paimon.table.sink.KeyAndBucketExtractor;
 
 import org.apache.flink.table.data.RowData;
 
-/** {@link AbstractChannelComputer} for {@link RowData}. */
-public class RowDataChannelComputer extends AbstractChannelComputer<RowData> {
+/** {@link ChannelComputer} for {@link RowData}. */
+public class RowDataChannelComputer implements ChannelComputer<RowData> {
 
-    public RowDataChannelComputer(
-            int numChannels, TableSchema schema, boolean shuffleByPartitionEnable) {
-        super(numChannels, new RowDataKeyAndBucketExtractor(schema), shuffleByPartitionEnable);
+    private static final long serialVersionUID = 1L;
+
+    private final TableSchema schema;
+    private final boolean shuffleByPartitionEnable;
+
+    private transient int numChannels;
+    private transient KeyAndBucketExtractor<RowData> extractor;
+
+    public RowDataChannelComputer(TableSchema schema, boolean shuffleByPartitionEnable) {
+        this.schema = schema;
+        this.shuffleByPartitionEnable = shuffleByPartitionEnable;
+    }
+
+    @Override
+    public void setup(int numChannels) {
+        this.numChannels = numChannels;
+        this.extractor = new RowDataKeyAndBucketExtractor(schema);
     }
 
     @Override
     public int channel(RowData record) {
-        return channelImpl(record);
+        extractor.setRecord(record);
+        return channel(extractor.partition(), extractor.bucket());
+    }
+
+    public int channel(BinaryRow partition, int bucket) {
+        if (shuffleByPartitionEnable) {
+            return ChannelComputer.channel(numChannels, bucket, partition);
+        } else {
+            return ChannelComputer.channel(numChannels, bucket);
```

Comment:
According #863, channel is selected by `bucket % numChannels`, and other keys, like partition, decide to choose the offset of the channel.

In this way, there is no need to have `shuffleByPartitionEnable`.